src/allmydata/test/eliotutil.py

```diff
@@ -174,10 +99,83 @@ def addCleanup(self, *a, **kw):
     def id(self):
         return self.case.id()
 
-    @eliot_logged_test
-    def run(self, result=None):
-        return self._run_tests_with_factory(
-            self.case,
-            self.handlers,
-            self.last_resort,
-        ).run(result)
+    def run(self, result):
+        """
+        Run the test case in the context of a distinct Eliot action.
+
+        The action will finish after the test is done.  It will note the name of
+        the test being run.
+
+        All messages emitted by the test will be validated.  They will still be
+        delivered to the global logger.
+        """
+        # The idea here is to decorate the test method itself so that all of
+        # the extra logic happens at the point where test/application logic is
+        # expected to be.  This `run` method is more like test infrastructure
+        # and things do not go well when we add too much extra behavior here.
+        # For example, exceptions raised here often just kill the whole
+        # runner.
+
+        # So, grab the test method.
+        name = self.case._testMethodName
+        original = getattr(self.case, name)
+        try:
+            # Patch in a decorated version of it.
+            setattr(self.case, name, with_logging(ensure_text(self.case.id()), original))
+            # Then use the rest of the machinery to run it.
+            return self._run_tests_with_factory(
+                self.case,
+                self.handlers,
+                self.last_resort,
+            ).run(result)
+        finally:
+            # Clean up the patching for idempotency or something.
+            delattr(self.case, name)
+
+
+def with_logging(
+        test_id,      # type: str
+        test_method,  # type: Callable
+):
+    """
+    Decorate a test method with additional log-related behaviors.
+
+    1. The test method will run in a distinct Eliot action.
+    2. Typed log messages will be validated.
+    3. Logged tracebacks will be added as errors.
+
+    :param test_id: The full identifier of the test being decorated.
+    :param test_method: The method itself.
+    """
+    @wraps(test_method)
+    def run_with_logging(*args, **kwargs):
+        validating_logger = _memory_logger()
+        original = swap_logger(None)
+        try:
+            swap_logger(_TwoLoggers(original, validating_logger))
+            with RUN_TEST(name=test_id):
+                try:
+                    return test_method(*args, **kwargs)
+                finally:
+                    check_for_errors(validating_logger)
+        finally:
+            swap_logger(original)
+    return run_with_logging
+
+
+class _TwoLoggers(object):
+    """
+    Log to two loggers.
+
```

Comment:
So "a logger" is just anything that has a `.write` method?

Reply:
Almost.  Added more docs to make this clear.